rdkit/Chem/__init__.py

```diff
@@ -42,6 +42,76 @@
   rdCoordGen.SetDefaultTemplateFileDir(templDir)
 
 
+# Github Issue #6208: boost::python iterators are slower than they should
+# (cause is that boost::python throws exceptions as actual C++ exceptions)
+class _GetAtomsIterator:
+
+  def __init__(self, mol):
+    self._mol = mol
+    self._pos = 0
+    self._size = mol.GetNumAtoms()
+
+  def __len__(self):
+    if self._mol.GetNumAtoms() != self._size:
+      raise RuntimeError('size changed during iteration')
+    return self._size
+
+  def __getitem__(self, i):
+    if i < 0 or i >= len(self):
+      raise IndexError('index out of range')
+    return self._mol.GetAtomWithIdx(i)
+
+  def __next__(self):
+    if self._pos >= len(self):
+      raise StopIteration
+
+    ret = self[self._pos]
+    self._pos += 1
+    return ret
+
+  def __iter__(self):
+    for i in range(0, len(self)):
+      self._pos = i
+      yield self[i]
+    self._pos = self._size
+
+
+class _GetBondsIterator:
+
+  def __init__(self, mol):
+    self._mol = mol
+    self._pos = 0
+    self._size = mol.GetNumBonds()
+
+  def __len__(self):
+    if self._mol.GetNumBonds() != self._size:
+      raise RuntimeError('size changed during iteration')
+    return self._size
+
+  def __getitem__(self, i):
+    if i < 0 or i >= len(self):
+      raise IndexError('index out of range')
+    return self._mol.GetBondWithIdx(i)
+
+  def __next__(self):
+    if self._pos >= len(self):
+      raise StopIteration
+
+    ret = self[self._pos]
+    self._pos += 1
+    return ret
+
+  def __iter__(self):
+    start = self._pos
```

Comment:
For simplicity it could `return self`, see https://wiki.python.org/moin/Iterator
 `__next__` should handle iteration then.

Reply:
No, that would change behavior from the previous implementation (it breaks tests).

Apparently, the iterator was implemented so that `__iterator__` restarts the iteration, while `__next__` continues from wherever the previous `__iterator__` iteration or `__next__` left it.